dek-registry-client/src/main/java/io/confluent/dekregistry/client/CachedDekRegistryClient.java

```diff
@@ -234,35 +291,88 @@ public void deleteKek(
+    deleteDek(DEFAULT_REQUEST_PROPERTIES, kekName, subject, algorithm, permanentDelete);
   }
 
   public void deleteDek(
-      Map<String, String> requestProperties, String kekName,
-      String subject, boolean permanentDelete)
+      Map<String, String> requestProperties, String kekName, String subject, DekFormat algorithm,
+      boolean permanentDelete)
       throws IOException, RestClientException {
-    deleteDek(requestProperties, kekName, subject, null, permanentDelete);
+    restService.deleteDek(requestProperties, kekName, subject, algorithm, permanentDelete);
+    dekCache.invalidateAll();
   }
 
   @Override
-  public void deleteDek(
-      String kekName, String subject, DekFormat algorithm, boolean permanentDelete)
+  public void deleteDekVersion(
+      String kekName, String subject, int version, DekFormat algorithm, boolean permanentDelete)
       throws IOException, RestClientException {
-    deleteDek(DEFAULT_REQUEST_PROPERTIES, kekName, subject, algorithm, permanentDelete);
+    deleteDekVersion(DEFAULT_REQUEST_PROPERTIES, kekName, subject, version,
+        algorithm, permanentDelete);
   }
 
-  public void deleteDek(
-      Map<String, String> requestProperties, String kekName, String subject, DekFormat algorithm,
-      boolean permanentDelete)
+  public void deleteDekVersion(
+      Map<String, String> requestProperties, String kekName, String subject, int version,
+      DekFormat algorithm, boolean permanentDelete)
       throws IOException, RestClientException {
-    restService.deleteDek(requestProperties, kekName, subject, algorithm, permanentDelete);
-    dekCache.invalidate(new DekId(kekName, subject, algorithm, permanentDelete));
+    restService.deleteDekVersion(requestProperties, kekName, subject, version,
+        algorithm, permanentDelete);
+    // Just invalidate all since the version can be represented many ways,
+    // such as null for first or -1 for latest
+    dekCache.invalidateAll();
+  }
+
+  @Override
+  public void undeleteKek(String kekName)
+      throws IOException, RestClientException {
+    undeleteKek(DEFAULT_REQUEST_PROPERTIES, kekName);
+  }
+
+  public void undeleteKek(
+      Map<String, String> requestProperties, String kekName)
+      throws IOException, RestClientException {
+    restService.undeleteKek(requestProperties, kekName);
+    kekCache.invalidate(new KekId(kekName, false));
+    kekCache.invalidate(new KekId(kekName, true));
+  }
+
+  @Override
+  public void undeleteDek(
+      String kekName, String subject, DekFormat algorithm)
+      throws IOException, RestClientException {
+    undeleteDek(DEFAULT_REQUEST_PROPERTIES, kekName, subject, algorithm);
+  }
+
+  public void undeleteDek(
+      Map<String, String> requestProperties, String kekName, String subject, DekFormat algorithm)
+      throws IOException, RestClientException {
+    restService.undeleteDek(requestProperties, kekName, subject, algorithm);
+    dekCache.invalidateAll();
```

Comment:
Why do we invalidate all? Does the dekCache hold deks for all keks? If so, why do we invalidate it all of we are only undeleting a certain dek?

Reply:
We don't expect the delete/undelete calls to be used that often.   Here it's easier to invalidate everything rather than try to remove all the DEKs for the given kekName, subject, and algorithm.  The DEKs will be re-added to the cache if they are retrieved again in the future.